net: lwm2m: DTLS session cache enable and queue mode update

jheiskan81 opened the pull request:
- Enabled DTLS session cache for support session resume.
- Fixed LwM2M queue mode for close connection and reconnect automatically.
- Re-connect will do Registration update before it send queued data.
- Coap Confirmation pending class allocation fix.
- Registration update failure trig full Registration. This will handle session resume failure.

Commit:
[nrf noup] net: lwm2m: DTLS session cache enable and queue mode update

Enabled DTLS session cache for support session resume.
Fixed LwM2M queue mode for close connection and reconnect automatically.
Re-connect will do Registration update before it send queued data.
Session resume is helping a case when NAT change address and cause less
network traffic.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

jheiskan81 commented:
@rlubos @SeppoTakalo @jarlamsa Please review.  I have tested this with Queue mode which was not working properly. But with this It keep registration valid at Coiote server and DTLS session resume is working also.

jheiskan81 commented:
static void socket_loop(void)
{
	int i, rc;
	int64_t timestamp;
	int32_t timeout, next_retransmit;

	while (1) {
		timestamp = k_uptime_get();
		timeout = lwm2m_engine_service(timestamp);

		/* wait for sockets */
		if (sock_nfds < 1) {
			k_msleep(timeout);
			continue;
		}
		
		That cause problems still. If I observe something We never reach notifcation if socket is closed :(

jheiskan81 commented:
@jheiskan81 and @rlubos I fix observation that I close socket and re open but not connected.  I added to connection re-open to socket_loop() when socket is not accept to send and it have pending tx. That solve problem and now observation re-connect client to server properly.

VeijoPesonen commented on include/net/lwm2m.h:
```diff
@@ -150,6 +150,13 @@ struct lwm2m_ctx {
 	bool use_dtls;
 
+	/**
+	 * Flag for indicate that Socket is connected.
```
"Flag to indicate that the socket is connected. With queue mode, this will tell if there is a need to reconnect" - or something similar.

jheiskan81 replied:
OK.  I looking for that. Thanks

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5369,6 +5374,43 @@ static int32_t lwm2m_engine_service(const int64_t timestamp)
 
+#if defined(CONFIG_LWM2M_QUEUE_MODE_ENABLED)
+
+int lwm2m_engine_close_socket_connection(struct lwm2m_ctx *client_ctx)
```
I just wonder, couldn't this just be a no-op function for UDP socket? I see little use for closing/reopening a socket in such case.

jheiskan81 replied:
Ok good point let me update function to more generic way.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5615,6 +5659,14 @@ static void socket_loop(void)
 			if (sock_fds[i].revents & POLLOUT) {
 				socket_send_message(sock_ctx[i]);
+			} else {
+#if defined(CONFIG_LWM2M_QUEUE_MODE_ENABLED)
+				int rc;
+				rc = lwm2m_rd_client_connection_resume(sock_ctx[i]);
+				if (rc) {
+					break;
+				}
+#endif
```
this won't work with native TLS, the DTLS behaviour is slightly different - i. e. when a socket does not have a valid DTLS session, it will initiate the handshake in the `sendto()` call. This effectively means, POLLOUT is set regardless of the handshake state.

Maybe we could just check the `connected` flag before send attempt?

jheiskan81 replied:
Let me re-think that. I see your point. I need to re-test this I have it some issue but I can re-investigated this.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5369,6 +5374,43 @@ static int32_t lwm2m_engine_service(const int64_t timestamp)
+#ifdef CONFIG_LWM2M_DTLS_SUPPORT
+	if (client_ctx->use_dtls) {
+		client_ctx->sock_fd =
+			socket(client_ctx->remote_addr.sa_family, SOCK_DGRAM, IPPROTO_DTLS_1_2);
```
Anyway, why is the socket reopened rightaway? It's a bit odd to me, that we're left with a misconfigured socket. Maybe, instead of having a separate `connected` flag it'd be enough to monitor `client_ctx->sock_fd`, and reopen/reconnect if it's set to -1?

jheiskan81 replied:
Let me test that I re-open at later on. I can test to remove connected bool and use sock_fd.

jheiskan81 commented:
@rlubos Hi. I update PR. Unfortunately I was not enable find solution for keep socket closed until we need re-connect. Problem is that resend and observation updates at socket_loop is not handled. 
I change place which do reconnect. Now Re-register and lwm2m_message_asynch_send() trig that. I update functionality that UDP socket is not close socket and re-open like you wish.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5707,6 +5797,16 @@ int lwm2m_socket_start(struct lwm2m_ctx *client_ctx)
 
+		int session_cache = TLS_SESSION_CACHE_ENABLED;
+
+		ret = setsockopt(client_ctx->sock_fd, SOL_TLS, TLS_SESSION_CACHE, &session_cache,
```
upstream session cache support is still on a PR (https://github.com/zephyrproject-rtos/zephyr/pull/43710)

jheiskan81 replied:
Interesting. It works in this repository. Re-connect just use 3 message vs 6 at full handshake.

jheiskan81 replied:
Ah I understand nrf91 is supporting that.

jheiskan81 commented:
@rlubos @VeijoPesonen @SeppoTakalo @jarlamsa Please review and comment's. AVSystem server not accept our Registration update if we build Queue client with cache. With Cache Client works very smoothly.
Changes from last are:
* Wake-up do re-connect and Registration update to server
* Client data is buffered and flushed to engine when Registration is ready
* Added KConfig for TLS Cache

jheiskan81 commented:
@rlubos @SeppoTakalo @jarlamsa @VeijoPesonen I simplify that Socket connect is handled only 1 place:
Re-registartion state. Message send trigger that state from RF_OFF state.

Commit:
[nrf noup] net: lwm2m: LwM2M RD client registration update failure fix

Jump to to full registration state if registration update fail.
Update keep already opened DTLS session.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

Commit:
[nrf noup] net: coap: Coap pending allocation fix

When sending 2 or more confirmable message before first one is
writed to socket all messages use same coap pending structure.
Now coap_pending_init() set data pointer which lock allocation by each call.

Signed-off-by: Juha Heiskanen <juha.heiskanen@nordicsemi.no>

jheiskan81 commented:
@rlubos @SeppoTakalo I have updated PR:
It include Coap module fix for allocate Coap pending structure after init.
Now Registration update failure not re-open DTLS it just do full registration.

jheiskan81 commented:
I would like step to make PR to zephyr Upstream soon.

VeijoPesonen commented on subsys/net/lib/coap/coap.c:
```diff
@@ -1147,7 +1147,7 @@ struct coap_pending *coap_pending_next_unused(
 
 	for (i = 0, p = pendings; i < len; i++, p++) {
-		if (p->timeout == 0) {
+		if (p->data == 0) {
```
You should write a comment why _data_ is used instead of _timeout_

jheiskan81 replied:
Ok Good point.

SeppoTakalo replied:
uh... I was just about to say that commenting this is quite unnecessary.. It should be said in the commit message why something was changed.
One the code, the change is not visible anymore so it does not deserve any comments how things have been in the past.

VeijoPesonen commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -861,6 +864,19 @@ static int sm_send_registration(bool send_obj_support_data,
 
+static void sm_handle_registration_update_failure(void)
+{
+	LOG_WRN("Registration Update fail -> trig full registration");
```
_trigger_

VeijoPesonen commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -186,6 +186,11 @@ config LWM2M_QUEUE_MODE_UPTIME
+config LWM2M_TLS_SESSION_CACHING
+	bool "Enable TLS session caching"
+	help
+	  Enable this only when Modem is supporting.
```
No need to mention the modem.

jheiskan81 commented:
@VeijoPesonen I fix your comment's.

rlubos commented on include/net/lwm2m.h:
```diff
@@ -150,6 +153,22 @@ struct lwm2m_ctx {
+
+	/**
+	 * Flag to indicate that the socket connection or registration is pending.
+	 * True value buffer Notification's and Send messages.
```
```suggestion
	 * True value buffer Notifications and Send messages.
```

rlubos commented on include/net/lwm2m.h:
```diff
@@ -162,6 +181,7 @@ struct lwm2m_ctx {
 	 */
 	bool bootstrap_mode;
 
+
```
Unneeded empty line.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1422,9 +1425,89 @@ int lwm2m_init_message(struct lwm2m_message *msg)
+int lwm2m_push_queued_buffers(struct lwm2m_ctx *client_ctx)
+{
+	client_ctx->buffer_client_messages = false;
+	while(1) {
```
Maybe use `!sys_slist_is_empty()` here?

jheiskan81 replied:
make sense.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1422,9 +1425,89 @@ int lwm2m_init_message(struct lwm2m_message *msg)
+
 int lwm2m_send_message_async(struct lwm2m_message *msg)
 {
+#if defined(CONFIG_LWM2M_QUEUE_MODE_ENABLED)
+#if defined CONFIG_LWM2M_RD_CLIENT_SUPPORT
```
I'd just use && operator and a single endif here.

jheiskan81 replied:
ok.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1422,9 +1425,89 @@ int lwm2m_init_message(struct lwm2m_message *msg)
+
+int lwm2m_information_interface_send(struct lwm2m_message *msg)
+{
+#if defined(CONFIG_LWM2M_QUEUE_MODE_ENABLED)
```
ditto

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -1148,6 +1182,33 @@ void lwm2m_rd_client_update(void)
+			 * Switch state for triggering a proper registration message
+			 * if CONFIG_LWM2M_TLS_SESSION_CACHING is false we force full
+			 * registration after Fully DTLS handshake
+			 */
```
The alignment seems off.

VeijoPesonen reviewed (approved):
Looks good to me

jarlamsa commented on include/net/lwm2m.h:
```diff
@@ -150,6 +153,22 @@ struct lwm2m_ctx {
 
+#if defined(CONFIG_LWM2M_QUEUE_MODE_ENABLED)
+	/**
+	 * Flag to indicate that the socket is connection is suspended.
```
nit: typo
```suggestion
	 * Flag to indicate that the socket connection is suspended.
```

jarlamsa commented on include/net/lwm2m.h:
```diff
@@ -150,6 +153,22 @@ struct lwm2m_ctx {
+	/**
+	 * Flag to indicate that the socket is connection is suspended.
+	 * With queue mode, this will tell if there is a need to reconnect.
+	 *
```
nit: Extra line

jarlamsa commented on include/net/lwm2m.h:
```diff
@@ -150,6 +153,22 @@ struct lwm2m_ctx {
+	/**
+	 * Flag to indicate that the socket connection or registration is pending.
+	 * True value buffer Notifications and Send messages.
+	 *
```
nit: Extra line

jarlamsa commented on include/net/lwm2m.h:
```diff
@@ -150,6 +153,22 @@ struct lwm2m_ctx {
+	bool connection_suspended;
+
+	/**
+	 * Flag to indicate that the socket connection or registration is pending.
```
```suggestion
	 * Flag to indicate that the client is buffering Notifications and Send messages. 
	 * This could be due to pending socket connection or pending registration.
```

jarlamsa commented on subsys/net/lib/lwm2m/Kconfig:
```diff
@@ -186,6 +186,11 @@ config LWM2M_QUEUE_MODE_UPTIME
+config LWM2M_TLS_SESSION_CACHING
+	bool "Enable TLS session caching"
+	help
+	  Enable this only when feature is supported.
```
```suggestion
	  Enable this only when feature is supported in Mbed TLS.
```

jheiskan81 replied:
I think we should not mention mbed TLS originally I have it "supported in modem"

jarlamsa replied:
It could be also just "in TLS library", currently it is a bit vague in my opinion.

jarlamsa reviewed (approved):
LGTM, just some small nits about the documentation in header.
Btw, you seem to be missing the sauce tags for the commits. The CI is nagging because of that.

jheiskan81 commented:
> LGTM, just some small nits about the documentation in header. Btw, you seem to be missing the sauce tags for the commits. The CI is nagging because of that.

Ah I will add those.

jheiskan81 commented:
@jarlamsa I fix all expected that mbed tls mention.

jheiskan81 commented:
@SeppoTakalo and @rlubos Do you have any change request or Can I merge this to WIP branch and start doing Upstream PR?

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -1422,9 +1425,85 @@ int lwm2m_init_message(struct lwm2m_message *msg)
+
+	if (client_ctx->connection_suspended) {
+		client_ctx->connection_suspended = false;
+		LOG_INF("Resume suspended connection");
```
```suggestion
		LOG_DBG("Resume suspended connection");
```

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -5707,6 +5839,18 @@ int lwm2m_socket_start(struct lwm2m_ctx *client_ctx)
 			goto error;
 		}
 
+		session_cache = IS_ENABLED(CONFIG_LWM2M_TLS_SESSION_CACHING) ?
+					   TLS_SESSION_CACHE_ENABLED :
+					   TLS_SESSION_CACHE_DISABLED;
+
+		ret = setsockopt(client_ctx->sock_fd, SOL_TLS, TLS_SESSION_CACHE, &session_cache,
+				 sizeof(session_cache));
+		if (ret < 0) {
+			ret = -errno;
+			LOG_ERR("Failed to set TLS_SESSION_CACHE option: %d", errno);
+			goto error;
+		}
```
```suggestion
		if (IS_ENABLED(CONFIG_LWM2M_TLS_SESSION_CACHING)) {
			int session_cache = TLS_SESSION_CACHE_ENABLED;

						ret = setsockopt(client_ctx->sock_fd, SOL_TLS, TLS_SESSION_CACHE, &session_cache,
					 sizeof(session_cache));
			if (ret < 0) {
				ret = -errno;
				LOG_ERR("Failed to set TLS_SESSION_CACHE option: %d", errno);
				goto error;
			}
		}
```

I would prefer you not to touch the value, unless enabled.

jheiskan81 replied:
OK I'm fine for this change.

jheiskan81 commented:
@SeppoTakalo Thanks for comment. I have fix those now.
